nexus/src/db/queries/external_ip.rs

```diff
@@ -887,4 +933,70 @@ mod tests {
 
         context.success().await;
     }
+
+    #[tokio::test]
+    async fn test_insert_external_ip_is_idempotent() {
+        let context =
+            TestContext::new("test_insert_external_ip_is_idempotent").await;
+
+        // Create an IP pool
+        let range = IpRange::try_from((
+            Ipv4Addr::new(10, 0, 0, 1),
+            Ipv4Addr::new(10, 0, 0, 3),
+        ))
+        .unwrap();
+        context.create_ip_pool("p0", range, None).await;
+
+        // Create one SNAT IP address.
+        let instance_id = Uuid::new_v4();
+        let id = Uuid::new_v4();
+        let project_id = Uuid::new_v4();
+        let ip = context
+            .db_datastore
+            .allocate_instance_snat_ip(
+                &context.opctx,
+                id,
+                project_id,
+                instance_id,
+            )
+            .await
+            .expect("Failed to allocate instance SNAT IP address");
+        assert_eq!(ip.kind, IpKind::SNat);
+        assert_eq!(ip.ip.ip(), range.first_address());
+        assert_eq!(ip.first_port.0, 0);
+        assert_eq!(
+            usize::from(ip.last_port.0),
+            super::NUM_SOURCE_NAT_PORTS - 1
+        );
+        assert_eq!(ip.project_id, project_id);
+
+        // Create a new IP, with the _same_ ID, and ensure we get back the same
+        // value.
+        let new_ip = context
+            .db_datastore
+            .allocate_instance_snat_ip(
+                &context.opctx,
+                id,
+                project_id,
+                instance_id,
+            )
+            .await
+            .expect("Failed to allocate instance SNAT IP address");
+
+        // Check identity, not equality. The timestamps will be updated.
+        assert_eq!(ip.id, new_ip.id);
+        assert_eq!(ip.name, new_ip.name);
+        assert_eq!(ip.description, new_ip.description);
+        assert!(ip.time_created < new_ip.time_created);
+        assert!(ip.time_modified < new_ip.time_modified);
```

Comment:
<= ?

Reply:
Yeah, I was on the fence about this. Strict inequality is fine for test though (1) it's not really guaranteed and (2) we don't care so much in real use cases. I'll update.

Reply:
Done in 5292bdab